### PR TITLE
[staging] git: 2.44.1 -> 2.45.1

### PR DESCRIPTION
**Pull request by JohnRTitor**
## Description of changes
- Supersedes #308186
- Changelog: https://github.com/git/git/blob/v2.45.1/Documentation/RelNotes/2.45.1.txt

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by FlorianFranzen** [comment]
Result of `nixpkgs-review pr 312211` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>git</li>
    <li>git.debug (git.debug.debug ,git.debug.doc)</li>
    <li>git.doc (git.doc.debug ,git.doc.doc)</li>
  </ul>
</details>

**Review by FlorianFranzen — APPROVED** [review]

**Review by Jayman2000 — CHANGES_REQUESTED** [review]
## Notes for the “git: 2.44.1 -> 2.45.1” commit

I was able to build and run these packages on NixOS 23.11 (x86-64):

- `git`
- `gitMinimal`
- `gitSVN`
- `gitFull`

I tried cloning a repo with each of them, and they all worked. For this commit, I only have one small suggestion. You wrote the commit message like this:

```
git: 2.44.1 -> 2.45.1
Fixes  CVE-2024-32002, CVE-2024-32004, CVE-2024-32020,
CVE-2024-32021 and CVE-2024-32465
https://github.com/git/git/blob/v2.45.1/Documentation/RelNotes/2.45.1.txt
```

I would add a blank line in between the subject and the body of the commit:

```
git: 2.44.1 -> 2.45.1

Fixes  CVE-2024-32002, CVE-2024-32004, CVE-2024-32020,
CVE-2024-32021 and CVE-2024-32465
https://github.com/git/git/blob/v2.45.1/Documentation/RelNotes/2.45.1.txt
```

## Notes for the “git: add passthru.update script” commit

The update script gives errors if `git` isn’t already up to date:

```console
$ git switch update-git 
Switched to branch 'update-git'
Your branch is up to date with 'pr-JohnRTitor/update-git'.
$ git revert de05a8c1cf94dc7f5eee0074ca91cb0e1d2481cd
Auto-merging pkgs/applications/version-management/git/default.nix
[update-git c18a58b71366] Revert "git: 2.44.1 -> 2.45.1"
 1 file changed, 2 insertions(+), 2 deletions(-)
$ nix-shell maintainers/scripts/update.nix --argstr package git

Going to be running update for following packages:
 - git-2.44.1

Press Enter key to continue...

Running update for:
 - git-2.44.1: UPDATING ...
 - git-2.44.1: ERROR

--- SHOWING ERROR LOG FOR git-2.44.1 ----------------------

grep: warning: stray \ before {
warning: could not open directory 'pkgs/applications/version-management/git-and-tools/git/': No such file or directory
fatal: pathspec '/home/jayman/Documents/Home/VC/Git/Partially mine/nixpkgs/repo/pkgs/applications/version-management/git-and-tools/git/default.nix' did not match any files


--- SHOWING ERROR LOG FOR git-2.44.1 ----------------------
The update script for git-2.44.1 failed with exit code 128
$ 
```

Despite the errors, it does successfully update `git`.

**Comment by JohnRTitor** [comment]
@ofborg build git
@ofborg build gitMinimal
@ofborg build gitSVN
@ofborg build gitFull

**Comment by JohnRTitor** [comment]
> I would add a blank line in between the subject and the body of the commit:

Done.

**Review by Jayman2000 — APPROVED** [review]
I still get errors when I run the update script, but it does work despite those errors. Considering the fact that this PR fixes security issues, my recommendation is to merge it as is. Once this PR gets merged, I’ll open an issue for the update script errors.

Edit: [It looks like someone beat me to it.](https://github.com/NixOS/nixpkgs/pull/314234)

**Review by SebTM — APPROVED** [review]
Result of `nixpkgs-review pr 312211` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>git</li>
    <li>git.debug (git.debug.debug ,git.debug.doc)</li>
    <li>git.doc (git.doc.debug ,git.doc.doc)</li>
  </ul>
</details>

**Comment by vcunat** [comment]
As the changelog says, this is not a security update.  We got the fixes in PR #311801 already.

File: pkgs/applications/version-management/git/default.nix [hunk 1]
```diff
@@ -233,51 +317,68 @@ stdenv.mkDerivation (finalAttrs: {
...
+                       --set GITPERLLIB "$out/${perlPackages.perl.libPrefix}:${
+                         perlPackages.makePerlPath (perlLibs ++ [ svn.out ])
+                       }" \
+                       --prefix PATH : "${svn.out}/bin" ''
+      else
+        ''
+          # replace git-svn by notification script
+                 notSupported $out/libexec/git-core/git-svn
+        ''
+    )
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/git/default.nix:334]: bruh, the formatter is just uglyfing the code

**Comment by SuperSandro2000** [comment]
I've dropped the formatting commit because I wanted to merge this PR and it just felt wrong adding so much noise to the file without any urgency. Doing the formatting in one go with a proper entry to `.git-blame-ignore-revs` is just the way to go instead of gradually introducing it to all kinds of files and creating noise and maybe even hitting some early adopter bugs which would create some extra noise in the history.

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**